crates/ra_mbe/src/mbe_expander.rs

```diff
@@ -441,29 +529,83 @@ fn expand_tt(
         }
         crate::TokenTree::Leaf(leaf) => match leaf {
             crate::Leaf::Ident(ident) => {
+                if ident.text == "macro_rules" {
```

Comment:
This special-casing feels sub-optimal... What would be long-term solution here?

I would guess that we should not want about unbound macro-variables at all during expansion, and instead report an error when parsing *after* expansion? 

Can we perhaps, instead of adding `macro_rules_state`, fix the `Err(err) => {` case bellow such that it just leaves the var intact?

Reply:
In fact, if we want, we can just check if `Bindings::inner` contains that `$var` key and than leave the var intact. But we lost the error reporting function of missing binding variable.

Reply:
I think this error reporting is misplaced: we should report unbound vars **after** expansion, precisely because of nested macros.

So, I prefer to regress tests that check for this error reporting to adding ad-hock tracking of macro scopes :)

Reply:
I think the long term solution here is an better mbe parser which can parse an inner mbe defintion rules, and collect all variables information within.

Reply:
But a good idea is to check how rustc reports this error!

Reply:
> I think the long term solution here is an better mbe parser which can parse an inner mbe defintion rules, and collect all variables information within.

I am not sure this could work: imagine something like this:

```
 macro_rules! foo {
  ($a:ident, $b:ident, $c:tt) => {
     my_proc_macro! bar {
         ($bi:ident) => {
             fn $bi() -> u8 {$c}
         }
     }
 }
```

where `my_proc_macro` is a procedural re-implmentation of macro rules. So, the parser wouldn't know thay `$bi` is a macro variable. 

We need to double check with rustc, but I think that passthrough of unbound variables is the only strategy that can work here (but I may be misunderstanding something!)

Reply:
Hm, this doesn't even need proc macros! Consider this example

```rust
macro_rules! identity {
    ($($tt:tt)*) => ($($tt)*)
}

macro_rules! foo {
    ($a:ident, $b:ident, $c:tt) => {
        identity! {
            macro_rules! bar {
                ($bi:ident) => { fn $bi() -> u8 {$c} }
            }
        }
    }
}
fn main() {
    foo!(a, b, 92);
    bar!(c);
    c();
}
```

(playground)[https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2911bee5ce577226190b88e80cfb4c99]

The parser cant' be smart enough to figure out what `identity` does, so it cant' understand that inner macro_rules is, in fact, macro_rules

Reply:
Um.. yes, seem to be you are correct.. :
https://github.com/rust-lang/rust/blob/767f59462663fbc55a69d39fc5e1f7f83b6fb37d/src/libsyntax/parse/parser.rs#L3270

This function is only called from `Parser::bump` ...Um...

Reply:
>This function is only called from Parser::bump ...Um...

Hehe, it's nice when an experiment confirms predictions of the theory! 

I haven't realized that this was the case when we originally implemented error detection here!

Reply:
Haha, and in fact i did implement a version which just by pass it :) 
Let me stash pop it and commit again.

Reply:
Fixed. :)